### PR TITLE
Added FadeIn data inside endpointregistry

### PR DESCRIPTION
**Pull request by RomanZavodskikh**
+ Added FadeIn data inside endpointregistry
+ Checking the data inside endpointegistry in *_test.go files.

### CODE REVIEW

File: routing/endpointregistry.go [hunk 1]
```diff
@@ -115,6 +122,15 @@ func (r *EndpointRegistry) DecInflightRequest(key string) {
 	e.inflightRequests--
 }
 
+func (r *EndpointRegistry) SetFadeInParams(key string, duration time.Duration, exponent float64) {
```
Comment (AlexanderYastrebov) [inline routing/endpointregistry.go:125]: Lets name it just `SetFadeIn`

File: routing/endpointregistry.go [hunk 2]
```diff
@@ -31,6 +34,10 @@ func (e *entry) InflightRequests() int64 {
 	return e.inflightRequests
 }
 
+func (e *entry) FadeInDuration() time.Duration {
```
Comment (AlexanderYastrebov) [inline routing/endpointregistry.go:37]: Do we need getter for exponent as well?

File: filters/fadein/fadein.go [hunk 3]
```diff
@@ -194,9 +199,14 @@ func (endpointCreated) Response(filters.FilterContext) {}
 
 // NewPostProcessor creates post-processor for maintaining the detection time of LB endpoints with fade-in
 // behavior.
-func NewPostProcessor() routing.PostProcessor {
+func NewPostProcessor(o Options) routing.PostProcessor {
```
Comment (AlexanderYastrebov) [inline filters/fadein/fadein.go:202]: Lets add `NewPostProcessorWithOptions` and then make
```go
func NewPostProcessor() routing.PostProcessor {
  return NewPostProcessorWithOptions(Options{})
}
```

File: routing/endpointregistry.go (hunk 2 above)
Comment (RomanZavodskikh) [reply routing/endpointregistry.go:37]: I can add it in the current PR, but we will need it only in the next PR (when loadbalancer algorithm will use it), so I decided to postpone this getter to the next PR.

File: filters/fadein/fadein.go [hunk 4]
```diff
@@ -228,13 +238,16 @@ func (p *postProcessor) Do(r []*routing.Route) []*routing.Route {
...
+			p.endpointRegistry.SetFadeInParams(ep.Host, ri.LBFadeInDuration, ri.LBFadeInExponent)
+
 			key := endpointKey(ep.Scheme, ep.Host)
 			detected := p.detected[key].when
 			if detected.IsZero() || endpointsCreated[key].After(detected) {
 				detected = now
 			}
 
 			ep.Detected = detected
+			p.endpointRegistry.SetDetectedTime(ep.Host, detected)
```
Comment (AlexanderYastrebov) [inline filters/fadein/fadein.go:250]: I think this should check if stored value is zero or endpointsCreated is after it (to override).

File: routing/endpointregistry.go [hunk 5]
```diff
@@ -14,10 +14,13 @@ const lastSeenTimeout = 1 * time.Minute
 type Metrics interface {
 	DetectedTime() time.Time
 	InflightRequests() int64
+	FadeInDuration() time.Duration
 }
 
 type entry struct {
 	detected         time.Time
+	fadeInDuration   time.Duration
```
Comment (RomanZavodskikh) [inline routing/endpointregistry.go:22]: In general, one endpoint could be used in several routes. So, it makes sense to create 1-to-many relation here somehow. The only thing that comes to my mind at the moment is map from routeId to struct with duration and exponent.

Comment (szuecs) [reply routing/endpointregistry.go:22]: Start simple, add metrics, measure, observe, optimize.

File: filters/fadein/fadein.go [hunk 6]
```diff
@@ -39,7 +43,8 @@ type (
 
 	postProcessor struct {
 		// "http://10.2.1.53:1234": {t0 60s t0-10s}
-		detected map[string]detectedFadeIn
+		detected         map[string]detectedFadeIn
+		endpointRegistry *routing.EndpointRegistry
```
Comment (szuecs) [inline filters/fadein/fadein.go:47]: why not have it as ptr?

File: filters/fadein/fadein.go [hunk 7]
```diff
@@ -192,14 +197,20 @@ func (endpointCreated) CreateFilter(args []interface{}) (filters.Filter, error)
...
+// the fade-in behavior.
+func NewPostProcessorWithOptions(o Options) routing.PostProcessor {
 	return &postProcessor{
-		detected: make(map[string]detectedFadeIn),
+		detected:         make(map[string]detectedFadeIn),
+		endpointRegistry: o.EndpointRegistry,
 	}
 }
 
+// Deprecated: use NewPostProcessorWithOptions
```
Comment (szuecs) [inline filters/fadein/fadein.go:209]: please use proper godoc (start with func name): 
```
NewPostProcessor *deprecated*: use NewPostProcessorWithOptions
```

Comment (RomanZavodskikh) [reply filters/fadein/fadein.go:209]: I think it's better not to use godoc starting with func name for deprecation to really work.
```
rzavodskikh@ZALANDO-39125:~/skipper$ staticcheck -checks "all,-ST1000,-ST1003,-ST1012,-ST1020,-ST1021" ./... # Start with "Deprecated"
skipper.go:1903:4: fadein.NewPostProcessor is deprecated: use NewPostProcessorWithOptions  (SA1019)
rzavodskikh@ZALANDO-39125:~/skipper$ staticcheck -checks "all,-ST1000,-ST1003,-ST1012,-ST1020,-ST1021" ./... # Start with "NewPostProcessor"
rzavodskikh@ZALANDO-39125:~/skipper$ 
```

File: proxy/fadeintesting_test.go [hunk 8]
```diff
@@ -270,7 +270,7 @@ func (p *fadeInProxy) addInstances(n int) {
 			DataClients:    []routing.DataClient{client},
 			PostProcessors: []routing.PostProcessor{
 				loadbalancer.NewAlgorithmProvider(),
-				fadein.NewPostProcessor(),
+				fadein.NewPostProcessorWithOptions(fadein.Options{}),
```
Comment (szuecs) [inline proxy/fadeintesting_test.go:273]: should not this also have the endpointsregistry passed?

Comment (RomanZavodskikh) [reply proxy/fadeintesting_test.go:273]: However, added endpointregistry here for the sake of uniformity, not to confuse anyone and to show other developers that it's a good practice to add it.

**Pull request closed without merging**